### PR TITLE
[Docs] APN correction proposal

### LINKED ISSUE #2643: Add APN Correction
[please attach design doc / any description]

### LINKED ISSUE #2650: [APN Correction]
- APN Correction: Virtual APN
  - Send different APN on PDPContextCreation based on user profile (ex: charging characristics: prepaid, postpaid)
- APN Correction: Based on IMSI prefix
  - Send different APN on PDPContextCreation based IMSI prefix

### PR DESCRIPTION
**Pull request by ymasmoudi**
Signed-off-by: YOUSSEF EL MASMOUDI <ymasmoudi@fb.com>

## Summary

APN Correction at the MME level. 

## Test Plan

NA

### CODE REVIEW

File: docs/readmes/proposals/p004_apn_correction.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+During session establishment, the phone may have an APN configured already.
+In this case, this APN will be used as it has a higher precedence over the
+one provided by the SIM card.
+
+This behavior requires a feature to override this value at the MME level
+based on IMSI prefixes prior to interrogating the HSS and creating the PDN
+session.
+
+This document describes the support of APN correction.
```
Comment (ulaskozat) [inline docs/readmes/proposals/p004_apn_correction.md]: The feature description is not clear to me. There is also repetition. 

Make it clear that this feature is for a federated set up where HSS is used and not used in standalone case.

File: docs/readmes/proposals/p004_apn_correction.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+
+In addition, a new function will be introduced to filter and return the overridden
+APN based on the MME NAS config.
+
+This function will be called upon reception of the ESM Information Request or PDN
+Connectivity Request.
+
+
+**NMS Change**
```
Comment (ulaskozat) [inline docs/readmes/proposals/p004_apn_correction.md]: What about orc8r changes?

File: docs/readmes/proposals/p004_apn_correction.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+3/ If no APN is included in the ESM Information Response or PDN Connectivity
+Request message, then the UE is connected to the default APN.
+
+
+The APN correction will be implemented upon reception of one of these two
+messages, PDN Connectivity Request or ESM Information Response. It will
+provide the possibility to filter IMSIs based on a prefix configured in the
+mme.yml or mconfig. 
+The APN is then overridden with the configured value.
```
Comment (ulaskozat) [inline docs/readmes/proposals/p004_apn_correction.md]: let's make it clear that the APN requested by UE is overwritten if `enable_apn_correction: true` for UEs with matching `imsi_prefix` filter specified as part of `apn_correction_map_list` with the value of key `apn_override`.

File: docs/readmes/proposals/p004_apn_correction.md (hunk 1 above)
Comment (ulaskozat) [reply docs/readmes/proposals/p004_apn_correction.md]: The description can be more specific:
" This feature enables the overriding of UE requested APN with a network specified APN via IMSI prefix based filtering. Up to 10 IMSI prefix filters and corresponding APNs to overwrite with can be defined."

File: docs/readmes/proposals/p004_apn_correction.md (hunk 2 above)
Comment (ymasmoudi) [reply docs/readmes/proposals/p004_apn_correction.md]: As per today swagger and nms doesn't support mme config.
So the only change will added to mconfig.proto for now.

**Review by themarwhal — APPROVED** [review]

**Pull request merged**